src/biokbase/narrative/jobs/specmanager.py

```diff
@@ -260,29 +271,35 @@ def _repr_html_(self):
 
         return Template(tmpl).render(usage=self.usage)
 
-    def __repr__(self):
+    def __repr__(self: "AppUsage") -> str:
+        """Generate a string representation of the instance."""
         return self.__str__()
 
-    def __str__(self):
-        s = "id: {}\nname: {}\nsubtitle: {}\nparameters (*required):\n-----------------------".format(
-            self.usage["id"], self.usage["name"], self.usage["subtitle"]
+    def __str__(self: "AppUsage") -> str:
+        """Generate a string representation of the instance."""
+        s = "\n".join(
+            [
+                f'id: {self.usage["id"]}',
+                f'name: {self.usage["name"]}',
+                f'subtitle: {self.usage["subtitle"]}',
+                "parameters (*required):",
+                "-----------------------",
+            ]
         )
-
         for p in self.usage["params"]:
             if not p.get("is_constant", False):
                 p_def = "\n{}{} - {}".format(
                     "*" if not p["optional"] else "", p["id"], p["type"]
                 )
                 if "allowed_types" in p:
                     p_def = (
-                        p_def
-                        + " - is a data object where the type is one of: {}".format(
-                            json.dumps(p["allowed_types"])
-                        )
+                        f"{p_def} - is a data object where the type is one of:"
+                        + json.dumps(p["allowed_types"])
                     )
+                # ??? indentation error?
                 if "allowed_values" in p:
-                    p_def = p_def + " - must be one of {}".format(
-                        json.dumps(p["allowed_values"])
+                    p_def = (
+                        f"{p_def} - must be one of {json.dumps(p['allowed_values'])}"
```

Comment:
not sure what is going on here - why there are two checks for the same thing. Suggests that some indentation might have been added? Who knows! Anyway, no user complaints, so leaving as-is.

Reply:
"allowed_types" vs "allowed_values", I think.

Reply:
Of course. Doh!